library/ssl_tls.c

```diff
@@ -5199,6 +5203,18 @@ int mbedtls_ssl_handshake_step( mbedtls_ssl_context *ssl )
     }
 #endif
 
+    if( ret != 0 )
+    {
+        int alert_ret;
+        alert_ret = mbedtls_ssl_handle_pending_alert( ssl );
```

Comment:
I did not think about it in my previous reviews but do we really want to overwrite the error code of the step processing function with the one of the function sending a potential pending alert ? The first one seems to convey more useful information than the latter.

Reply:
I don't know if the (non-MPS) code achieves that, but I think the intent was that we remember the original error code and return it after sending the alert has succeeded. If the alert fails with `WANT_WRITE`, we'd like the user to call again to finish sending the alert, and then observe the original error.

Reply:
It is directly copy from non-MPS code.

Reply:
[`MPS`](https://github.com/hannestschofenig/mbedtls/blob/ae6624780fba2d9256072b73338c67b84f9b9793/library/ssl_msg.c#L6205) and [`non-MPS`](https://github.com/hannestschofenig/mbedtls/blob/ae6624780fba2d9256072b73338c67b84f9b9793/library/ssl_msg.c#L6241) are different. 

I think it should be modified according to `MPS` version

Reply:
As Hanno said, in case `mbedtls_ssl_handle_pending_alert()` returns `WANT_WRITE` I think we want to return that. Otherwise I'm honestly not sure which is best to return.

There aren't that many reasons why `mbedtls_ssl_send_alert_message()` would fail: the underlying TCP connection has been closed, the crypto accelerator is in an error state... if that happens, then it's probably just as good to return that as the original error, since it means your connection is dead anyway.

Thinking while writing: there's at least one case where the original error may be more informative: if the peer sends an alert then immediately closes the TCP socket, then it's probably slightly better to return `MBEDTLS_ERR_SSL_FATAL_ALERT_MESSAGE` than `MBEDTLS_ERR_SSL_CONN_EOF` though tbh they're quite similar: the peer wasn't happy and stopped talking to us. Either way, the next step is to try again with debug messages enabled, because even if there was an alert, just the return value isn't going to tell you which alert it was.

The other reason `mbedtls_ssl_handle_pending_alert()` would return non-zero is if `mbedtls_ssl_send_alert_message()` succeeded, and then it returns `ssl->alert_reason` which as the comment points out is very likely to be the same as as `ret` anyway.

Overall, we probably should avoid overthinking which error message is more informative (as long as they're correct or course!) for now, for two reasons: (1) error messages carry limited information and you'll usually have to rely on debug messages to understand the problem anyway, and (2) we _know_ we are going to revisit this when switching to MPS, so this is one of the rare cases where we can leave something for later and know that "later" doesn't mean "never".

Reply:
Thanks for the inputs. Let's stick with the error codes of the  last version of this PR.